### ISSUE TITLE
Allow to add a call signature to the Mapped Type OR to remove all `Function.prototype` methods

### ISSUE THREAD

**Issue by anurbol** [microsoft/TypeScript#27575]
## Search Terms

add a call signature to the Mapped Type; remove all `Function.prototype` methods

## Suggestion

Being able to create a callable type, but without Function.prototype methods. 

## Use Cases

Imagine one creates some API that exposes a function that is (because function is an object) also have some custom properties. This function is not supposed to be `call`-ed, `apply`-ed, `bind`-ed, etc, in fact all the `Function.prototype` methods are removed from it (e.g. with setPrototypeOf..null). However IDE will still suggest the `Function.prototype` methods.

*Desired behavior:*
```ts
let callableObject = () => 'foo'
Object.setPrototypeOf(callableObject, null)
callableObject.bar = 'baz'

callableObject() // 'foo'
callableObject // {bar: 'baz'}
callableObject.call // <---------- error
```
*Current behavior:*
```ts
let callableObject = () => 'foo'
Object.setPrototypeOf(callableObject, null)
callableObject.bar = 'baz'

callableObject() // 'foo'
callableObject // {bar: 'baz'}
callableObject.call // <---------- OK
```

*What I tried:*
```ts
type ExcludeFunctionPrototypeMethods<T extends () => any> = {
    [K in Exclude<keyof T, keyof Function>]: T[K]
}
// the type above "lacks a call signature" and I have no Idea how to add it there.
```

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by RyanCavanaugh** [comment]
```ts
type Unfunction<T> = T & { call: null; apply: null; bind: null; }

function getUnfunction<T extends Function>(callableObject: T): Unfunction<T> {
    Object.setPrototypeOf(callableObject, null);
    return callableObject as Unfunction<T>;
}



let callableObject = getUnfunction(() => 'foo');

callableObject() // 'foo'
callableObject // {bar: 'baz'}
callableObject.call() // <---------- error
```

**Comment by RyanCavanaugh** [comment]
You could also use `never` in place of `null`

**Comment by anurbol** [comment]
Hi, Ryan! Thank you for the response! 

The solution you provided can let user know that `.call` has null/never type. However it still appears in the list of other, API properties (`.bar`). I realize there may not be any solution to solve it, but if so can you please answer two questions that arised from this:

1. I always wondered: why do properties with type `never` still appear in IntelliSense? I ask this here because I guess the TypeScript Language Server sends it to the IntelliSense.
2. Is Mapped Types not able having call signature is a design decision or it's not just implemented yet (because of few use cases)?

**Comment by trickeyd** [comment]
The problem with this method is that all the prototype props still exist on the type, just as 'never'. I feel there should be a way to explicitly 'mask' types out completely so that they don't appear in the type at all - for the sake of a clean api.

Been talking about it here:
https://stackoverflow.com/questions/59700434/typescript-omitting-a-function-prototype-from-a-type-extending-any?noredirect=1#comment105570732_59700434 

this is kinda linked too:
https://github.com/Microsoft/TypeScript/issues/27575

**Comment by trickeyd** [comment]
I'd also like to remove the function prototype props:

https://stackoverflow.com/questions/59700434/typescript-omitting-a-function-prototype-from-a-type-extending-any?noredirect=1#comment105570732_59700434

https://github.com/Microsoft/TypeScript/issues/27575

**Comment by fwh1990** [comment]
I agree with @anurbol . Sometimes we have custom property on function like `fn.foo = 123`, and we want user see this property from IDE tips, however, so many unexpected properties like `apply | call | bind | arguments ...` are shown, in other words `fn.foo` is hidden from them, user is very hard to find custom things in a second.

**Comment by nth-commit** [comment]
+1

Trying to create a concise function builder, and having the interesting methods polluted by the function prototype.

**Comment by zachkirsch** [comment]
+1 would love to exclude extra these function properties to make the IDE experience better

**Comment by atablash** [comment]
We could really use a feature where **mapped types would not strip call and construct signatures** or there would be a new mapping type that would perform mapping on call and/or construct signatures instead of (or in addition to) regular property mapping.

For **callables**, this would be very useful for implementing any sort of callable interfaces.

For **newables**, perhaps more common, you could perform property mappings on constructors with additional properties (classes with static fields) - without loosing the construct signature.

**Comment by Krombik** [comment]
I found a workaround:

```ts
declare class SomeName {
  static someField: any;
}

type Identity<T> = T;

interface SomeName extends Identity<typeof SomeName> {
  (...args: any[]): any;
}

const someName: SomeName
```

this solution allows `someName` to be called and the `someField` property to be first in the suggestion list

**Comment by geoffreytools** [comment]
I like @Krombik's workaround and I think this should be the default display order for properties on callable objects.

I feel that it is generally a terrible idea to remove properties methods from the function prototype just to get cleaner tooltips. Unless the implementer truly has a good reason to prescribe how users should call their functions.

**Comment by rotu** [comment]
@RyanCavanaugh this breaks in TS 4.3.5, where it deduces the type of `callableObject` as `never` (I think it's actually broken in 3.9.7, but there is no visible error on calling `never`). [playground link](https://www.typescriptlang.org/play?strictFunctionTypes=false&target=99&jsx=0&ts=4.3.5#code/C4TwDgpgBAqgdgMwK5wMbAJYHs4B4AqAfFALxT5QBkUA3lKgIYA2TAXFHEiwNxQNhgmIdpx5QARhjgATEVya8AvgChlyNJhxQA5hGDx16bHgoQAHsAgyAzlABiKIzkIAKRiwbimEAPLiAVhDo7PgAlOwGjpomxDTKUAlQfoHoAHTWegAKAE5YwHngvghuzEye3slBwAA0HPKh3PGJ2XpI2XD0peW+AVV8tpEaxgSEjSqq3sCdHl49KVNkuvqIUcYuLqGkxADkCFhY2w2q7mWzlegbUAD0V1C7+9vKJ93nUze04gzZ7NufAF7bFTPM69NInS7vXAAWhhsLhUAg2Vy2SAA)

Is there an existing issue for this?

> ```ts
> type Unfunction<T> = T & { call: null; apply: null; bind: null; }
> 
> function getUnfunction<T extends Function>(callableObject: T): Unfunction<T> {
>     Object.setPrototypeOf(callableObject, null);
>     return callableObject as Unfunction<T>;
> }
> 
> 
> 
> let callableObject = getUnfunction(() => 'foo');
> 
> callableObject() // 'foo'
> callableObject // {bar: 'baz'}
> callableObject.call() // <---------- error
> ```

**Comment by jcalz** [comment]
> We could really use a feature where **mapped types would not strip call and construct signatures** or there would be a new mapping type that would perform mapping on call and/or construct signatures instead of (or in addition to) regular property mapping.
> 
> For **callables**, this would be very useful for implementing any sort of callable interfaces.
> 
> For **newables**, perhaps more common, you could perform property mappings on constructors with additional properties (classes with static fields) - without loosing the construct signature.

#29261 seems to be the place for that

**Comment by pfdgithub** [comment]
try
```
type OmitFuncProto<
  Func extends Function,
  ProtoKeys extends string | number | symbol = Function extends Record<infer K, any> ? K : never,
> = Omit<Func, ProtoKeys>;
```

**Comment by ahzvenol** [comment]
+1, I use Proxy to map function properties to another object to create my reactive store. In this case, Function.prototype doesn't exist and is of no use.

**Comment by guyutongxue** [comment]
Run into this issue, but here is a small tips: mark all keys of `Function` to `@deprecated`:
```ts
type NotFunctionPrototype = {
  /** @deprecated This object do not have function prototype */
  apply: never;
  /** @deprecated This object do not have function prototype */
  bind: never;
  /** @deprecated This object do not have function prototype */
  call: never;
  /** @deprecated This object do not have function prototype */
  arguments: never;
  /** @deprecated This object do not have function prototype */
  caller: never;
  /** @deprecated This object do not have function prototype */
  prototype: never;
  /** @deprecated This object do not have function prototype */
  toString: never;
  /** @deprecated This object do not have function prototype */
  length: never;
  /** @deprecated This object do not have function prototype */
  name: never;
};
type Foo = ((...args: any[]) => number) & NotFunctionPrototype
declare let foo: Foo;
foo.
//.  ^~ sorts all Function.prototype methods at end of suggestion list
```

And when user typing `.` after the object, they won't be facing annoying Function.prototype at their IDE suggestion list :)

<img width="1010" height="622" alt="Image" src="https://github.com/user-attachments/assets/25f29650-e36d-44bf-a670-9ac0dbd3ffd7" />
